### ISSUE TITLE
ReverseDiffAdjoint gives zero gradients after update

### ISSUE THREAD

**Issue by andrschl** [SciML/DiffEqSensitivity.jl#378]
After a update the following code for continuous loss with ReverseDiffAdjoint fails and gives zero gradients. Not sure whether it is due to ReverseDiff or DiffEqSensitivity. The same works with ForwardDiffSensitivity or with ReverseDiff v1.4.3 and DiffEqSensitivity v6.34.0.
```
model = Chain(Dense(2,2,tanh), Dense(2,2))
p,re = Flux.destructure(model)
g = (u,p)->re(p)(u)
function ODE_func!(du,u,p,t)
    x = u[1:2]
    du.=vcat(g(x,p), sum(abs2,g(x,p))*ones(1))
end
alg = Tsit5()
prob_cont = ODEProblem(ODE_func!, u0_cont, t_span, p=p)
dldp = Zygote.gradient(p) do p
    loss = Array(solve(prob_cont, alg, u0=u0_cont, p=p, saveat=[t[end]],sensealg=ReverseDiffAdjoint()))[3,1]
    return loss
end
```

**Comment by ChrisRackauckas** [comment]
I think this might be an isolated case of some of the `vcat` issues I've been seeing in ReverseDiff.

@mohamed82008

**Comment by andrschl** [comment]
Thanks! Ah yeah there was a "Fix single number (v/h)cat" PR from ReverseDiff v1.4.3 -> v1.4.4. Any ideas how to work around this?

**Comment by andrschl** [comment]
What's strange is that `ReverseDiff.gradient(u->ODE_func(u,p,t), u)` and `ReverseDiff.gradient(p->ODE_func(u,p,t), p)` work but they do not work for the inplace definition of `ODE_func!` as above.

**Comment by ChrisRackauckas** [comment]
It's because you need to turn the TrackedArray into Array{TrackedReal} to handle the mutation.

**Comment by devmotion** [comment]
I am not sure how made up the MWE is but regardless of AD I assume it would be better to implement it without `vcat` and to not recompute `g(x, p)` - e.g., something like
```julia
function ODE_func!(du, u, p, t)
    x = u[1:2]
    y = g(x, p)
	du[1:(end-1)] .= y
	du[end] = sum(abs2, y)
end
```

**Comment by andrschl** [comment]
> I am not sure how made up the MWE is but regardless of AD I assume it would be better to implement it without vcat and to not recompute g(x, p)

Yeah the example was completely made up. But still your assignment for `du` makes much more sense than with `vcat` and it works :D

> It's because you need to turn the TrackedArray into Array{TrackedReal} to handle the mutation.

Ok, so maybe it was the combination of this workaround and some vcat issue with ReverseDiff? I was just afraid because this stopped working after the update. But I guess I can close this when it is an isolated issue which can easily be avoided.
